### PR TITLE
Adding UpdateActivityOptions, first iteration

### PR DESCRIPTION
**Pull request by ychebotarev**
## What changed?

First iteration for UpdateActivityOptions.
Contains most of the logic for UpdateActivityOptions, working with proto, updating options, usage of  updateMask,  unit tests, etc.
Functional tests will come next.
To reviewers - 
* check the main logic in updateactivityoptions/api.go.
* check if anything fundamental is missing.

Known missing parts - reroute activity to the new TaskQueue is TaskQueue name was changes, activity is scheduled but not started yet.
Knows issues with proto - I was not able to make protorefrect work with "messages as a pointers", IsValid (should return false if field is a pointer and nil) is just not working. As the result code is a bit less general. Though more readable.

### CODE REVIEW

File: service/history/api/updateactivityoptions/api_test.go [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+		mask     *fieldmaskpb.FieldMask
+	}{
+		{
+			name:     "full mix - CamelCase",
+			options:  allOptions,
+			ai:       &persistencespb.ActivityInfo{},
+			expected: fullActivityInfo,
+			mask: &fieldmaskpb.FieldMask{
+				Paths: []string{
+					"TaskQueue.Name",
```
Comment (alexshtin) [inline service/history/api/updateactivityoptions/api_test.go:97]: Do you think we want to support all cases combination? What is best practices here? I would start with names from proto files, but I don't have strong opinion here.

File: service/history/tasks/activity_retry_timer.go [hunk 2]
```diff
@@ -41,6 +41,7 @@ type (
 		EventID             int64
 		Version             int64
 		Attempt             int32
+		Stamp               int32
```
Comment (alexshtin) [inline service/history/tasks/activity_retry_timer.go:44]: You need to add [serializer](https://github.com/temporalio/temporal/blob/d34367b20235490a632b49a0f096114bf1634fa5/common/persistence/serialization/task_serializer.go#L721)/deserializer also.

File: service/history/timer_queue_active_task_executor.go [hunk 3]
```diff
@@ -310,6 +277,66 @@ Loop:
...
+func (t *timerQueueActiveTaskExecutor) processTimeoutTask(
+	mutableState workflow.MutableState,
+	timerSequenceID workflow.TimerSequenceID,
+	ai *persistencespb.ActivityInfo,
+	updateMutableState *bool,
+	scheduleWorkflowTask *bool,
+) (retError error) {
+
+	if timerSequenceID.Attempt < ai.Attempt {
+		// 2. timerSequenceID.attempt < activityInfo.Attempt
```
Comment (alexshtin) [inline service/history/timer_queue_active_task_executor.go:289]: 2?

File: service/history/timer_queue_active_task_executor.go
Comment (alexshtin) [inline service/history/timer_queue_active_task_executor.go:1]: Changes in this file look like more or less separated refactoring. Extract another PR just for it?

File: service/history/timer_queue_active_task_executor.go [hunk 4]
```diff
@@ -310,6 +277,66 @@ Loop:
 	return t.updateWorkflowExecution(ctx, weContext, mutableState, scheduleWorkflowTask)
 }
 
+func (t *timerQueueActiveTaskExecutor) processTimeoutTask(
+	mutableState workflow.MutableState,
+	timerSequenceID workflow.TimerSequenceID,
+	ai *persistencespb.ActivityInfo,
+	updateMutableState *bool,
+	scheduleWorkflowTask *bool,
```
Comment (alexshtin) [inline service/history/timer_queue_active_task_executor.go:285]: I think these should be return values.

File: service/history/timer_queue_active_task_executor.go [hunk 5]
```diff
@@ -310,6 +277,66 @@ Loop:
...
+) (retError error) {
```
Comment (alexshtin) [inline service/history/timer_queue_active_task_executor.go:286]: And this shouldn't be named return value:
```suggestion
) error {
```

File: service/history/workflow/activity.go [hunk 6]
```diff
@@ -57,7 +69,7 @@ func nextRetryDelayFrom(failure *failurepb.Failure) *time.Duration {
 	return delay
 }
 
-func updateActivityInfoForRetries(
+func UpdateActivityInfoForRetries(
```
Comment (alexshtin) [inline service/history/workflow/activity.go:72]: doesn't seem like you need to export this.

File: service/history/workflow/activity.go [hunk 7]
```diff
@@ -76,3 +88,43 @@ func updateActivityInfoForRetries(
 
 	return ai
 }
+
+func NextBackoffInterval(
```
Comment (alexshtin) [inline service/history/workflow/activity.go:92]: can't find usages of this either.

File: service/history/workflow/mutable_state_impl.go [hunk 8]
```diff
@@ -4914,6 +4914,10 @@ func (ms *MutableStateImpl) RetryActivity(
 	return enumspb.RETRY_STATE_IN_PROGRESS, nil
 }
 
+func (ms *MutableStateImpl) GenerateActivityRetryTask(ai *persistencespb.ActivityInfo) error {
```
Comment (alexshtin) [inline service/history/workflow/mutable_state_impl.go:4917]: And this :-)

File: service/history/api/updateactivityoptions/api.go [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+	// move forward activity version
+	ai.Stamp += 1
+
+	// invalidate timers
+	ai.TimerTaskStatus = workflow.TimerTaskStatusNone
+
+	// regenerate retry tasks
+	if workflow.GetActivityState(ai) == enumspb.PENDING_ACTIVITY_STATE_SCHEDULED {
+		// two options - it can be in backoff, or waiting to be started
+		now := shardContext.GetTimeSource().Now().In(time.UTC)
```
[the remark this answers is not in the document]
Comment (ychebotarev) [reply service/history/api/updateactivityoptions/api.go:84]: done

File: service/history/api/updateactivityoptions/api.go [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+		},
+	}
+
+	return &api.UpdateWorkflowAction{
+		Noop:               false,
+		CreateWorkflowTask: false,
+	}, nil
+}
+
+func Invoke(
```
[the remark this answers is not in the document]
Comment (ychebotarev) [reply service/history/api/updateactivityoptions/api.go:120]: done

File: service/history/api/updateactivityoptions/api.go [hunk 11]
```diff
@@ -0,0 +1,223 @@
...
+
+func updateActivityOptions(
+	shardContext shard.Context,
+	mutableState workflow.MutableState,
+	request *historyservice.UpdateActivityOptionsRequest,
+	response *historyservice.UpdateActivityOptionsResponse,
+) (*api.UpdateWorkflowAction, error) {
+	if !mutableState.IsWorkflowExecutionRunning() {
+		return nil, consts.ErrWorkflowCompleted
+	}
```
[the remark this answers is not in the document]
Comment (ychebotarev) [reply service/history/api/updateactivityoptions/api.go:56]: removed

File: service/history/api/updateactivityoptions/api.go [hunk 12]
```diff
@@ -0,0 +1,223 @@
...
+		return serviceerror.NewInvalidArgument("UpdateMask is nil")
+	}
+
+	updateFields := make(map[string]struct{})
+
+	for _, path := range mask.Paths {
+		pathParts := util.ConvertPathToCamel(path)
+		jsonPath := strings.Join(pathParts, ".")
+		updateFields[jsonPath] = struct{}{}
+	}
```
[the remark this answers is not in the document]
Comment (ychebotarev) [reply service/history/api/updateactivityoptions/api.go:171]: extracted to common/proto

File: service/history/api/updateactivityoptions/api_test.go (hunk 1 above)
Comment (ychebotarev) [reply service/history/api/updateactivityoptions/api_test.go:97]: in short - yes. (we also support "taskQueue.name" btw). This is because both are legit ways, and we may be in a situation when we don't control how those field masks are created

File: service/history/timer_queue_active_task_executor.go (hunk 3 above)
Comment (ychebotarev) [reply service/history/timer_queue_active_task_executor.go:289]: leftover. removed

File: service/history/timer_queue_active_task_executor.go
Comment (ychebotarev) [reply service/history/timer_queue_active_task_executor.go:1]: it is a mix of refactoring/adding new functionality, so lets keep it this way.

File: service/history/timer_queue_active_task_executor.go (hunk 4 above)
Comment (ychebotarev) [reply service/history/timer_queue_active_task_executor.go:285]: sigh. it is ugly both ways to me...

File: common/util/proto.go [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+package util
+
+import (
+	"strings"
+
+	"golang.org/x/text/cases"
+	"golang.org/x/text/language"
+)
+
+func ConvertPathToCamel(input string) []string {
```
[the remark this answers is not in the document]
Comment (bergundy) [reply common/util/proto.go:32]: Can also use this: https://github.com/iancoleman/strcase

File: service/history/timer_queue_active_task_executor.go (hunk 4 above)
Comment (ychebotarev) [reply service/history/timer_queue_active_task_executor.go:285]: actually - I tried - return values are even more ugly. 

it will be either named return values (which is , to my best knowledge, something go community  currently recommend not to do), or plain `return true, false, nil`, with is error prone and not readable.

Best thing will be to extract all of that to a separate structure. It started good....but then we have a random emitTimeoutMetricScopeWithNamespaceTag call. 
Which cannot be extracted (used in few other places), cannot be converted into separate function, and also need to be wrapped into some sort of interface because "testing"

File: service/history/timer_queue_active_task_executor.go (hunk 5 above)
Comment (ychebotarev) [reply service/history/timer_queue_active_task_executor.go:286]: removed

File: service/history/api/updateactivityoptions/api.go [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+
+	// update activity options
+	err := applyActivityOptions(ai, activityOptions, updateRequest.GetUpdateMask())
+	if err != nil {
+		return nil, err
+	}
+
+	// move forward activity version
+	ai.Stamp += 1
```
[the remark this answers is not in the document]
Comment (ychebotarev) [reply service/history/api/updateactivityoptions/api.go:76]: done

File: service/history/workflow/activity.go (hunk 6 above)
Comment (ychebotarev) [reply service/history/workflow/activity.go:72]: Yeah. In my first implementation I was using it. Will revert this change.

File: service/history/workflow/activity.go (hunk 7 above)
Comment (ychebotarev) [reply service/history/workflow/activity.go:92]: yeah, it was an attempt to refactor a bit. Will remove it

File: service/history/workflow/mutable_state_impl.go (hunk 8 above)
Comment (ychebotarev) [reply service/history/workflow/mutable_state_impl.go:4917]: removed

File: service/history/tasks/activity_retry_timer.go (hunk 2 above)
Comment (ychebotarev) [reply service/history/tasks/activity_retry_timer.go:44]: probably done...please check if anything else is needed.

File: service/history/api/updateactivityoptions/api.go [hunk 15]
```diff
@@ -0,0 +1,223 @@
...
+		ai.RetryMaximumInterval = ao.RetryPolicy.MaximumInterval
+	}
+	if _, ok := updateFields["retryPolicy.maximumAttempts"]; ok {
+		if ao.RetryPolicy == nil {
+			return serviceerror.NewInvalidArgument("RetryPolicy is not provided")
+		}
+		ai.RetryMaximumAttempts = ao.RetryPolicy.MaximumAttempts
+	}
+
+	return nil
```
[the remark this answers is not in the document]
Comment (ychebotarev) [reply service/history/api/updateactivityoptions/api.go:209]: I refactor the code, extracted comman validator, and use it to validate ActivityOptions

File: service/history/api/updateactivityoptions/api.go [hunk 16]
```diff
@@ -0,0 +1,266 @@
...
+			}
+		}
+	}
+
+	// fill the response
+	response.ActivityOptions = mergeInto
+
+	return &api.UpdateWorkflowAction{
+		Noop:               false,
+		CreateWorkflowTask: false,
```
Comment (bergundy) [inline service/history/api/updateactivityoptions/api.go:169]: nit: these are defaults, no need to specify them explicitly.

File: service/history/api/updateactivityoptions/api.go [hunk 17]
```diff
@@ -0,0 +1,266 @@
...
+		return nil, err
+	}
+
+	// validate the updated options
+	err = validateActivityOptions(validator, request.NamespaceId, ai, mergeInto)
+	if err != nil {
+		return nil, err
+	}
+
+	// move forward activity version
```
Comment (bergundy) [inline service/history/api/updateactivityoptions/api.go:142]: Where's the assignment from `mergeInto` to `ai`?

File: service/history/timer_queue_active_task_executor.go [hunk 18]
```diff
@@ -310,6 +277,65 @@ Loop:
 	return t.updateWorkflowExecution(ctx, weContext, mutableState, scheduleWorkflowTask)
 }
 
+func (t *timerQueueActiveTaskExecutor) processTimeoutTask(
```
Comment (bergundy) [inline service/history/timer_queue_active_task_executor.go:280]:
```suggestion
func (t *timerQueueActiveTaskExecutor) processSingleActivityTimeoutTask(
```

File: service/history/timer_queue_active_task_executor.go (hunk 4 above)
Comment (bergundy) [reply service/history/timer_queue_active_task_executor.go:285]: Agree this should be extracted to return values. You can create a struct to make it more readable.

File: service/history/timer_queue_active_task_executor.go [hunk 19]
```diff
@@ -310,6 +277,65 @@ Loop:
...
+	mutableState workflow.MutableState,
+	timerSequenceID workflow.TimerSequenceID,
+	ai *persistencespb.ActivityInfo,
+	updateMutableState *bool,
+	scheduleWorkflowTask *bool,
+) error {
+
+	if timerSequenceID.Attempt < ai.Attempt {
+		//  retry could update activity attempt, should not timeouts new attempt
```
Comment (bergundy) [inline service/history/timer_queue_active_task_executor.go:289]:
```suggestion
		//  The RetryActivity call below could update activity attempt, in which case we do not want to apply a timeout for the previous attempt.
```

File: service/history/timer_queue_active_task_executor.go [hunk 20]
```diff
@@ -310,6 +277,65 @@ Loop:
...
+		return nil
+	}
+
+	if timerSequenceID.Stamp != ai.Stamp {
+		// this timer event is from an old stamp. In this case we ignore the event.
+		*updateMutableState = true
```
Comment (bergundy) [inline service/history/timer_queue_active_task_executor.go:295]: Why update mutable state here?

File: service/history/timer_queue_active_task_executor.go [hunk 21]
```diff
@@ -310,6 +277,65 @@ Loop:
...
+	updateMutableState *bool,
+	scheduleWorkflowTask *bool,
+) error {
+
+	if timerSequenceID.Attempt < ai.Attempt {
+		//  retry could update activity attempt, should not timeouts new attempt
+		return nil
+	}
+
+	if timerSequenceID.Stamp != ai.Stamp {
```
Comment (bergundy) [inline service/history/timer_queue_active_task_executor.go:293]: How can this happen? Isn't the timer sequence generated in the same transaction that this function is called?

File: service/history/timer_queue_active_task_executor.go [hunk 22]
```diff
@@ -310,6 +277,65 @@ Loop:
...
+		// this timer event is from an old stamp. In this case we ignore the event.
+		*updateMutableState = true
+		return nil
+	}
+
+	failureMsg := fmt.Sprintf("activity %v timeout", timerSequenceID.TimerType.String())
+	timeoutFailure := failure.NewTimeoutFailure(failureMsg, timerSequenceID.TimerType)
+	retryState, err := mutableState.RetryActivity(ai, timeoutFailure)
```
Comment (bergundy) [inline service/history/timer_queue_active_task_executor.go:301]: I think you only want to update mutable state if this is called, right?

File: service/history/timer_queue_active_task_executor.go (hunk 21 above)
Comment (bergundy) [reply service/history/timer_queue_active_task_executor.go:293]: I think you want this in `executeActivityRetryTimerTask` and `transferQueueActiveTaskExecutor.processActivityTask`.

File: service/history/api/updateactivityoptions/api.go [hunk 23]
```diff
@@ -0,0 +1,266 @@
...
+		return nil, err
+	}
+
+	validator := api.NewCommandAttrValidator(
+		shardContext.GetNamespaceRegistry(),
+		shardContext.GetConfig(),
+		nil,
+	)
+
+	response := &historyservice.UpdateActivityOptionsResponse{}
```
[the remark this answers is not in the document]
Comment (ychebotarev) [reply service/history/api/updateactivityoptions/api.go:62]: response is used as a parameter for updateActivityOptions

File: service/history/api/updateactivityoptions/api.go [hunk 24]
```diff
@@ -0,0 +1,266 @@
...
+	updateRequest := request.GetUpdateRequest()
+	mergeFrom := updateRequest.GetActivityOptions()
+	activityId := updateRequest.GetActivityId()
+
+	ai, activityFound := mutableState.GetActivityByActivityID(activityId)
+
+	if !activityFound {
+		// Looks like ActivityTask already completed as a result of another call.
+		// It is OK to drop the task at this point.
+		return nil, consts.ErrActivityTaskNotFound
```
[the remark this answers is not in the document]
Comment (ychebotarev) [reply service/history/api/updateactivityoptions/api.go:106]: I will add different error.

File: service/history/api/updateactivityoptions/api.go [hunk 25]
```diff
@@ -0,0 +1,266 @@
...
+	// invalidate timers
+	ai.TimerTaskStatus = workflow.TimerTaskStatusNone
+
+	// regenerate retry tasks
+	if workflow.GetActivityState(ai) == enumspb.PENDING_ACTIVITY_STATE_SCHEDULED {
+		// two options - it can be in backoff, or waiting to be started
+		now := shardContext.GetTimeSource().Now()
+		if now.After(ai.ScheduledTime.AsTime()) {
+			// activity is past its scheduled time and ready to be started
+			// we don't really need to do generate timer tasks, it should be done in closeTransaction
```
[the remark this answers is not in the document]
Comment (ychebotarev) [reply service/history/api/updateactivityoptions/api.go:154]: I can, prefer not to. I did it for readability.
I specifically want to add a comment that there is nothing to do in this case. 
I will change the order with empty `else`, compiler will optimize it.

File: service/history/api/updateactivityoptions/api.go [hunk 26]
```diff
@@ -0,0 +1,266 @@
...
+) error {
+	if mergeInto == nil {
+		return serviceerror.NewInvalidArgument("MergeInto argument is not provided")
+	}
+	if mergeFrom == nil {
+		return serviceerror.NewInvalidArgument("mergeFrom argument is not provided")
+	}
+	if updateFields == nil {
+		return serviceerror.NewInvalidArgument("updateFields argument is not provided")
+	}
```
[the remark this answers is not in the document]
Comment (ychebotarev) [reply service/history/api/updateactivityoptions/api.go:186]: yeah, will move to the upper level.

File: service/history/api/updateactivityoptions/api.go (hunk 16 above)
Comment (ychebotarev) [reply service/history/api/updateactivityoptions/api.go:169]: This is how it is in the rest of the code, which I think is for readability. It is better to be explicit with what function is returning. Compiler will take care of it.

File: service/history/api/updateactivityoptions/api.go (hunk 17 above)
Comment (ychebotarev) [reply service/history/api/updateactivityoptions/api.go:142]: ah, in my local changes...
will be in the next rev.

File: service/history/timer_queue_active_task_executor.go (hunk 18 above)
Comment (ychebotarev) [reply service/history/timer_queue_active_task_executor.go:280]: ok.

File: service/history/timer_queue_active_task_executor.go (hunk 21 above)
Comment (bergundy) [reply service/history/timer_queue_active_task_executor.go:293]: Discussed offline, only needed in `executeActivityRetryTimerTask` for now. Pausing the activity is a different story.

Comment (ychebotarev) [reply service/history/timer_queue_active_task_executor.go:293]: 1. add task.Stamp check earlier.
2.  updated executeActivityRetryTimerTask function - add stamp check.

File: service/history/timer_queue_active_task_executor.go (hunk 20 above)
Comment (ychebotarev) [reply service/history/timer_queue_active_task_executor.go:295]: this code is gone

File: service/history/timer_queue_active_task_executor.go (hunk 19 above)
Comment (ychebotarev) [reply service/history/timer_queue_active_task_executor.go:289]: done

File: service/history/timer_queue_active_task_executor.go (hunk 4 above)
Comment (ychebotarev) [reply service/history/timer_queue_active_task_executor.go:285]: ok :)
still ugly though...

File: service/history/api/command_attr_validator.go
[the remark this answers is not in the document]
Comment (ychebotarev) [reply service/history/api/command_attr_validator.go:1]: yeah, some name changes but mostly copy/paste

File: service/history/api/updateactivityoptions/api.go [hunk 27]
```diff
@@ -0,0 +1,266 @@
...
+	request *historyservice.UpdateActivityOptionsRequest,
+	shardContext shard.Context,
+	workflowConsistencyChecker api.WorkflowConsistencyChecker,
+) (resp *historyservice.UpdateActivityOptionsResponse, retError error) {
+	_, err := api.GetActiveNamespace(shardContext, namespace.ID(request.GetNamespaceId()))
+	if err != nil {
+		return nil, err
+	}
+
+	validator := api.NewCommandAttrValidator(
```
[the remark this answers is not in the document]
Comment (ychebotarev) [reply service/history/api/updateactivityoptions/api.go:56]: Technically we shouldn't make it static, because both config and registry are coming from from shardContext, and shardContext can be different for different instances of this call.

So if for whatever reason config/registry is changed, we will be working with outdated config/registry.

That said - we do internally know that both config and registry are practically singletons...
So I would rather think about making them static instead of passing them all the way to where they are needed.

File: service/history/api/updateactivityoptions/api.go [hunk 28]
```diff
@@ -0,0 +1,266 @@
...
+	"go.temporal.io/server/service/history/workflow"
+)
+
+func Invoke(
+	ctx context.Context,
+	request *historyservice.UpdateActivityOptionsRequest,
+	shardContext shard.Context,
+	workflowConsistencyChecker api.WorkflowConsistencyChecker,
+) (resp *historyservice.UpdateActivityOptionsResponse, retError error) {
+	_, err := api.GetActiveNamespace(shardContext, namespace.ID(request.GetNamespaceId()))
```
[the remark this answers is not in the document]
Comment (ychebotarev) [reply service/history/api/updateactivityoptions/api.go:51]: this line is here mostly I found the same line in many other api calls, without much comments. 

I think it is not needed. We usually wrap incoming request into internal proto which has "NamespaceId", and when we do that we are doing the same call (in this case):
`namespaceId, err := wh.namespaceRegistry.GetNamespaceID(namespace.Name(request.GetNamespace()))`

I will remove it. Probably also not needed in many other places.

File: common/util/proto.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (ychebotarev) [reply common/util/proto.go:33]: Lets leave it as is for now, and revisit once Alex will come with those helpers

File: service/history/api/respondworkflowtaskcompleted/command_checker.go
Comment (alexshtin) [inline service/history/api/respondworkflowtaskcompleted/command_checker.go:1]: Can you `git mv` entire file and then copy back size checker? Or move size checker too. Then history is preserved.

File: service/history/api/updateactivityoptions/api.go [hunk 29]
```diff
@@ -0,0 +1,264 @@
...
+		TaskQueue:              ao.TaskQueue,
+		ScheduleToCloseTimeout: ao.ScheduleToCloseTimeout,
+		ScheduleToStartTimeout: ao.ScheduleToStartTimeout,
+		StartToCloseTimeout:    ao.StartToCloseTimeout,
+		HeartbeatTimeout:       ao.HeartbeatTimeout,
+		ActivityId:             ai.ActivityId,
+		ActivityType:           ai.ActivityType,
+	}
+
+	_, err := validator.ValidateActivityScheduleAttributes(namespace.ID(namespaceID), attributes, nil)
```
Comment (alexshtin) [inline service/history/api/updateactivityoptions/api.go:262]: If you look closer you will see that `Validate...` not only validate but also fixes some fields in command attributes. You need to copy them back.

File: service/history/consts/const.go [hunk 30]
```diff
@@ -67,6 +67,8 @@ var (
 	ErrTransitionHistoryDisabled = serviceerror.NewFailedPrecondition("Transition history disabled")
 	// ErrActivityTaskNotFound is the error to indicate activity task could be duplicate and activity already completed
 	ErrActivityTaskNotFound = serviceerror.NewNotFound("invalid activityID or activity already timed out or invoking workflow is completed")
+	// ErrActivityNotFound is the error to indicate that there is no pending activity with this ID
+	ErrActivityNotFound = serviceerror.NewNotFound("Can't find pending activity with such ID. Invalid activityID or activity already completed")
```
Comment (alexshtin) [inline service/history/consts/const.go:71]: I think I reached some dzen in `ID` vs `Id` discussion. I can accept both.

**Review by alexshtin — APPROVED** [review]

**Comment by alexshtin** [comment]
Functional tests are definitely need but I am ok if you do this in separate PR.

File: service/history/consts/const.go (hunk 30 above)
Comment (ychebotarev) [reply service/history/consts/const.go:71]: monkey see monkey do. In all other places it is "activityID" (one of them right above)

File: service/history/api/respondworkflowtaskcompleted/command_checker.go
Comment (ychebotarev) [reply service/history/api/respondworkflowtaskcompleted/command_checker.go:1]: discussed. I will rename (git mv) command_checker to workflow_size_checker

File: service/history/api/command_attr_validator.go
[the remark this answers is not in the document]
Comment (ychebotarev) [reply service/history/api/command_attr_validator.go:1]: discussed offline. there are plenty of files there, refactoring should be in a separate PR.

File: service/history/api/updateactivityoptions/api.go (hunk 29 above)
Comment (ychebotarev) [reply service/history/api/updateactivityoptions/api.go:262]: good catch. I did notice that...but totally forgot.

File: service/history/timer_queue_active_task_executor.go [hunk 31]
```diff
@@ -228,6 +230,18 @@ func (t *timerQueueActiveTaskExecutor) executeActivityTimeoutTask(
...
+	if !ok {
+		// if activity is not found, the timer is invalid.
+		return nil
+	}
+
+	if task.Stamp != ai.Stamp {
+		// if this task is invalid - we want to generate new activity timeout task
+		// it will be done in closeTransactionPrepareTasks call
+		return weContext.UpdateWorkflowExecutionAsActive(ctx, t.shardContext)
+	}
```
[the remark this answers is not in the document]
Comment (ychebotarev) [reply service/history/timer_queue_active_task_executor.go:244]: Without it we can have duplicate retries.

File: service/history/tasks/activity_task_timer.go [hunk 32]
```diff
@@ -42,6 +42,7 @@ type (
 		TimeoutType         enumspb.TimeoutType
 		EventID             int64
 		Attempt             int32
+		Stamp               int32
```
[the remark this answers is not in the document]
Comment (ychebotarev) [reply service/history/tasks/activity_task_timer.go:45]: it is needed to avoid duplicate timer tasks (activity moved from 1 min to 10 min, and then back).

File: service/history/api/respondworkflowtaskcompleted/workflow_task_completed_handler.go [hunk 33]
```diff
@@ -1246,16 +1247,15 @@ func (handler *workflowTaskCompletedHandler) handleCommandModifyWorkflowProperti
 	// get namespace name
 	executionInfo := handler.mutableState.GetExecutionInfo()
 	namespaceID := namespace.ID(executionInfo.NamespaceId)
-	namespaceEntry, err := handler.namespaceRegistry.GetNamespaceByID(namespaceID)
+	_, err := handler.namespaceRegistry.GetNamespaceByID(namespaceID)
 	if err != nil {
 		return nil, serviceerror.NewUnavailable(fmt.Sprintf("Unable to get namespace for namespaceID: %v.", namespaceID))
 	}
```
[the remark this answers is not in the document]
Comment (ychebotarev) [reply service/history/api/respondworkflowtaskcompleted/workflow_task_completed_handler.go:1253]: I removed unused argument. I don't know this code enough to remove it.

File: service/history/api/updateactivityoptions/api.go [hunk 34]
```diff
@@ -0,0 +1,280 @@
...
+	ai.TimerTaskStatus = workflow.TimerTaskStatusNone
+
+	// regenerate retry tasks
+	if workflow.GetActivityState(ai) == enumspb.PENDING_ACTIVITY_STATE_SCHEDULED {
+		// two options - it can be in backoff, or waiting to be started
+		now := shardContext.GetTimeSource().Now()
+
+		// if activity is past its scheduled time and ready to be started
+		// we don't really need to do generate timer tasks, it should be done in closeTransaction
+		if now.Before(ai.ScheduledTime.AsTime()) {
```
[the remark this answers is not in the document]
Comment (ychebotarev) [reply service/history/api/updateactivityoptions/api.go:181]: discussed offline. I will change the approach in the next iteration

**Comment by yycptt** [comment]
Please do not land PRs with unresolved comments/discussions.
